ConsoleView: Recursively get steps for all stages (and child stages)

timbrown5 opened the pull request:
<!-- Please describe your pull request here. -->

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md  in your own repository 
-->

I realised the current code only loops through the top level of stages when trying to find steps. This fixes the code so  it gets the steps for each stage (but recursive down to each child).


Fixes #22 
> The FrontEnd doesn't seem to call steps API for the lowest level stages (they are always empty).

timbrown5 commented:
The code in this PR is currently broken. I have uplaoded it in the hope that someone can see what I am doing wrong.

I am trying to record the stage we are interested in in enclosingNode:
https://github.com/timbrown5/pipeline-graph-view-plugin/blob/addStepsToAllLevels/src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java#L149

What I thiink is happening is that this is too aggressive and get set to a child of the parallel block when we want it to he the parallel block itself (meaning the key in the map is wrong).

Here is where we add the step to the stageStepMap:
https://github.com/timbrown5/pipeline-graph-view-plugin/blob/addStepsToAllLevels/src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java#L188

timja commented:
Is there a good example pipeline to reproduce with?

timbrown5 commented:
> Is there a good example pipeline to reproduce with?
The screenshot above is using the pipeline from this comment (under 'Nested').
https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/15#issuecomment-865661205

timja commented:
This simple pipeline is now logged out of order:
https://github.com/timja-org/junit-attachments-test/blob/master/Jenkinsfile

timja commented:
Looks like the list of steps is just in reverse (ideally needs some test coverage)

timbrown5 commented:
> Looks like the list of steps is just in reverse (ideally needs some test coverage), not sure if we need to reverse, adjust parsing, or sort by node id?

That's a good point. I think if I push the steps to a stack and then empty the stack when I get to a start node it will fix that issue (and hopefully the other issue).

timbrown5 commented:
- Updated the visitor to use a stack and which fixed the ordering. 
- Updated the API so it only creates one instance of the StepVisitor - it was to aid debugging, but should actually make it more efficient too.
- Update StepVisitor to store a list of pending blocks (if we hit and end, we add to pending, when we hit start we pop from pending). This allowed me to only add steps to a stage if it's parent wasn't a parallel stage - which looked like it's resolved the issue.

![Screenshot 2021-07-25 at 10 44 23](https://user-images.githubusercontent.com/4447764/126894891-f38bd321-22be-4d28-9d7e-e91c6fe30eda.png)
 
Here is the log from both of the visitors:
<details>
  <summary>Full log</summary>
2021-07-25 09:43:13.302+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Non-Parallel Stage' with id '6' is type 'STAGE'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Parallel Stage' with id '11' is type 'STAGE'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Branch A' with id '15' is type 'PARALLEL'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Branch B' with id '16' is type 'PARALLEL'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Branch C' with id '17' is type 'PARALLEL'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Skipped stage' with id '59' is type 'STAGE'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Nested 1' with id '25' is type 'STAGE'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Nested 2' with id '41' is type 'STAGE'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Nested 3' with id '47' is type 'STAGE'.
2021-07-25 09:43:13.502+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 59 - true - false - false .
2021-07-25 09:43:13.503+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 59 - true - false - false .
2021-07-25 09:43:13.503+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.503+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 0.
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 11 - true - false - false .
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchEnd: Node ID: 17 - false - true - false .
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 23 - true - false - false .
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 47 - true - false - false .
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 49 to stack.
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 48 to stack.
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 47 - true - false - false .
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '48' to '47'.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '49' to '47'.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 41 - true - false - false .
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 43 to stack.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 42 to stack.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 41 - true - false - false .
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '42' to '41'.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '43' to '41'.
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 25 - true - false - false .
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 37 to stack.
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 36 to stack.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 25 - true - false - false .
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '36' to '25'.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '37' to '25'.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 23 - true - false - false .
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Not pushing steps to stage in handleChunkDone as parent is parallel block.
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Node ID: 17 - false - true - false .
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Pushing steps to stage in parallelBranchStart.
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 0.
2021-07-25 09:43:13.512+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchEnd: Node ID: 16 - false - true - false .
2021-07-25 09:43:13.516+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 21 - true - false - false .
2021-07-25 09:43:13.516+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 30 to stack.
2021-07-25 09:43:13.516+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 29 to stack.
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 21 - true - false - false .
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Not pushing steps to stage in handleChunkDone as parent is parallel block.
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Node ID: 16 - false - true - false .
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Pushing steps to stage in parallelBranchStart.
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '29' to '16'.
2021-07-25 09:43:13.518+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '30' to '16'.
2021-07-25 09:43:13.518+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.518+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchEnd: Node ID: 15 - false - true - false .
2021-07-25 09:43:13.527+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 19 - true - false - false .
2021-07-25 09:43:13.527+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 27 to stack.
2021-07-25 09:43:13.528+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 26 to stack.
2021-07-25 09:43:13.529+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 19 - true - false - false .
2021-07-25 09:43:13.529+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Not pushing steps to stage in handleChunkDone as parent is parallel block.
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Node ID: 15 - false - true - false .
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Pushing steps to stage in parallelBranchStart.
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '26' to '15'.
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '27' to '15'.
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 11 - true - false - false .
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 0.
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 6 - true - false - false .
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 7 to stack.
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 6 - true - false - false .
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '7' to '6'.
2021-07-25 09:43:13.533+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 1.
</details>

timja commented on src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:
```diff
@@ -56,12 +62,13 @@ public void getSteps(StaplerRequest req, StaplerResponse rsp) throws IOException
-            LOGGER.log(Level.FINE, "Steps: '" + mapper.writeValueAsString(stepApi.getSteps()) + "'.");
-            rsp.getWriter().append(mapper.writeValueAsString(stepApi.getSteps()));
+            PipelineStepList steps = stepApi.getSteps(nodeId);
+            LOGGER.log(Level.FINE, "Steps: '" + mapper.writeValueAsString(steps) + "'.");
```
should be wrapped in either a supplier or an `if LOGGER.isLoggable` pattern so the object isn't converted to a string when not needed

timja commented on src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApi.java:
```diff
@@ -22,41 +22,45 @@
+    private static Object mutex = new Object();
 
-    public PipelineStepApi(WorkflowRun run, String nodeId) throws java.io.IOException {
+    public PipelineStepApi(WorkflowRun run, String nodeId) {
```
node id is assigned but never accessed is there a reason for that?

timbrown5 replied:
An artefact of the was I was trying to get the steps before.

timja commented on src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineGraphApi.java:
```diff
@@ -120,11 +127,85 @@ public PipelineGraph createGraph() {
+
+        Map<Integer, List<Integer>> stageToChildrenMap = new HashMap<>();
+
+        List<Integer> stageIds = new ArrayList<>();
```
this list doesn't appear to be used? (other than adding to it)

timja commented on src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java:
```diff
@@ -56,14 +56,20 @@
     private boolean stageStepsCollectionCompleted = false;
 
     private boolean inStageScope;
+    private boolean inParallelBranchScope;
```
unused?

timja commented on src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java:
```diff
@@ -56,14 +56,20 @@
 
     private boolean inStageScope;
+    private boolean inParallelBranchScope;
+    private boolean parentnIsParallelBranch;
```
unused and typo on name?

timja reviewed (changes requested):
I've done some code cleanup, can you take a look at the rest of my comments? 

Is there anything else outstanding on this. I've re-tested and the issues I mentioned before are fixed

timbrown5 commented:
Sorry managed to make my changes when in a detached head, so pushed the one instance of the branch.

timbrown5 commented:
> Is there anything else outstanding on this. I've re-tested and the issues I mentioned before are fixed

I would like to add a test for the StepAPI before this is landed. Your cleanup inadvertently broke the parsing (I did the same thing last night when trying to clean up locally). Adding a test should stop this from happening again - I started this on a different branch before so should be able to revive it.

timja commented:
> > Is there anything else outstanding on this. I've re-tested and the issues I mentioned before are fixed
> 
> I would like to add a test for the StepAPI before this is landed. Your cleanup inadvertently broke the parsing (I did the same thing last night when trying to clean up locally). Adding a test should stop this from happening again - I started this on a different branch before so should be able to revive it.

interesting, didn't think I changed any logic, just places where my IDE warned the condition could not be true

timbrown5 commented:
> interesting, didn't think I changed any logic, just places where my IDE warned the condition could not be true
Spotbugs was warning about this
`pendingBlocks.pop()` which got removed (but is needed to track which stage we should be adding the steps to.

timbrown5 commented:
> I would like to add a test for the StepAPI before this is landed. Your cleanup inadvertently broke the parsing (I did the same thing last night when trying to clean up locally). Adding a test should stop this from happening again - I started this on a different branch before so should be able to revive it.

I have added tests to check that the PipelineGraphAPI.createTree and PipelineStepsAPI.getSteps work as expected in our simple use case.

timja commented on src/test/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApiTest.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    @Test
+    public void complexParallelParallelBranchesHaveCorrectSteps() throws Exception {
```
```suggestion
    public void complexParallelBranchesHaveCorrectSteps() throws Exception {
```
